{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## IMPORT PACKAGES\n",
    "\n",
    "import os,gzip,itertools,csv,re,sys,io, Bio\n",
    "import os.path\n",
    "import urllib.request\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import urllib\n",
    "import csv\n",
    "import fastaparser as fp\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio import Entrez\n",
    "from ftplib import FTP\n",
    "import warnings\n",
    "from pathlib import Path\n",
    "import difflib \n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=DeprecationWarning)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## READING IN SUMMARY DATA\n",
    "\n",
    "dat = pd.read_csv('ensemble_metazoa.csv')\n",
    "\n",
    "\n",
    "#with open('ensemble_metazoa.csv', \"r\", newline='') as csvfile:\n",
    "#    read = csv.reader(csvfile, delimiter=',')\n",
    "#    for row in read:\n",
    "#        if row[2] != \"Crustacea\" or \"Hemiptera\" or \"Diptera\" or \"Coleoptera\" or \"Hymenoptera\" or \"Lepidoptera\" or \"Chelicerata\" or \"Phthiraptera\" or \"Isoptera\":\n",
    "            \n",
    "#dat\n",
    "#dat.Classification\n",
    "\n",
    "insecta = dat[(dat.Accession.isin([\"GCA_000142985.2\", \"GCA_002204515.1\", \"GCA_001444175.1\", \"GCA_003254395.2\", \"GCA_000143395.2\", \"GCA_000775305.1\", \"GCA_000188095.2\", \"GCA_000214255.1\", \"GCA_000151625.1\", \"GCA_000648675.1\", \"GCA_000235995.2\", \"GCA_000313835.1\", \"GCA_000208615.1\", \"GCA_002892825.2\", \"GCA_003675905.1\", \"GCA_001187945.1\", \"GCA_000716385.1\", \"GCA_000371365.1\", \"GCA_009193385.2\", \"GCA_000006295.1\", \"GCA_000181055.3\", \"GCA_000828355.1\", \"GCA_000188075.2\", \"GCA_000611955.2\", \"GCA_001015335.1\", \"GCA_002237135.1\", \"GCA_000239435.1\", \"GCA_007210705.1\", \"GCA_000002335.3\", \"GCA_002443255.1\", \"GCA_000696155.1\"]))]\n",
    "\n",
    "insecta_df = insecta[[\"Name\", \"Classification\", \"Taxon ID\", \"Assembly\", \"Accession\"]]\n",
    "\n",
    "#removing duplicates\n",
    "norm_insecta_df = insecta_df.drop(insecta_df.index[13])\n",
    "\n",
    "#norm_insecta_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## CODE TO PARSE FASTA FILES (Did not use this)\n",
    "# this is code which will parse FASTA files\n",
    "# define what a header looks like in FASTA format\n",
    "def isheader(line):\n",
    "    return line[0] == '>'\n",
    "\n",
    "def aspairs(f):\n",
    "    seq_id = ''\n",
    "    sequence = ''\n",
    "    for header,group in itertools.groupby(f, isheader):\n",
    "        if header:\n",
    "            line = next(group)\n",
    "            seq_id = line[1:].split()[0]\n",
    "        else:\n",
    "            sequence = ''.join(line.strip() for line in group)\n",
    "            yield seq_id, sequence\n",
    "            \n",
    "            \n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## GFF3 DATA PARSING\n",
    "\n",
    "# Get names\n",
    "GFFdirnames = os.listdir('GFF3s')\n",
    "\n",
    "#Loop through files\n",
    "GFFdir = os.scandir('GFF3s')\n",
    "#print(GFFdirnames)\n",
    "\n",
    "## Obtain genome stats\n",
    "GFFdat = []\n",
    "for f in GFFdir:\n",
    "    with gzip.open(f, \"rt\") as fh:\n",
    "        mygff = csv.reader(fh, delimiter=\"\\t\")\n",
    "        count = 0\n",
    "        gene_len = 0\n",
    "        CDS_count = 0\n",
    "        CDS_len = 0\n",
    "        exon_count = 0\n",
    "        exon_len = 0\n",
    "        intron_count = 0\n",
    "        intron_len = 0\n",
    "        genome_size=0\n",
    "        for row in mygff:\n",
    "            if row[0].startswith(\"#\"):\n",
    "                continue\n",
    "            #print(row)\n",
    "            if row[2] == \"gene\":\n",
    "                count += 1\n",
    "                total_gene_len = int(row[4])-int(row[3])\n",
    "                gene_len += total_gene_len\n",
    "            if row[2] == \"CDS\":\n",
    "                CDS_count += 1\n",
    "                total_CDS_len =int(row[4])-int(row[3])\n",
    "                CDS_len +=  total_CDS_len\n",
    "            if row[2] == \"exon\":\n",
    "                exon_count += 1\n",
    "                total_exon_len =int(row[4])-int(row[3])\n",
    "                exon_len +=  total_exon_len\n",
    "                    \n",
    "        mylist=list([Path(f).stem, count, gene_len, exon_count, exon_len, CDS_count, CDS_len])\n",
    "\n",
    "    GFFdat.append(mylist)\n",
    "#print(GFFdat)\n",
    "           "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## CREATE DF FROM GFF DATA\n",
    "\n",
    "GFFdf = pd.DataFrame(GFFdat, columns = [\"Name\",\"Num_of_Genes\",\"Gene_Length\",\"Num_of_Exons\",\"Exon_Length\",\"Num_of_CDS\",\"CDS_Length\"]).sort_values(\"Name\") \n",
    "GFFdf\n",
    "\n",
    "#removing duplicates\n",
    "norm1_GFFdf=GFFdf.drop(GFFdf.index[13])\n",
    "norm_GFFdf=norm1_GFFdf.drop(GFFdf.index[27])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "##OBTAIN GENOME SIZE\n",
    "\n",
    "gensizedir = os.scandir('ncbi-genomes-2020-12-17')\n",
    "\n",
    "names = []\n",
    "size = []\n",
    "for t in gensizedir:\n",
    "    with open(t, \"r\") as txt:\n",
    "        for x in txt:\n",
    "            #print(x)\n",
    "            if \"Organism name\" in x:\n",
    "                names.append(x)\n",
    "            if x.startswith('#'):\n",
    "                continue\n",
    "            elif 'Primary Assembly' in x:\n",
    "                continue\n",
    "            elif 'assembled-molecule' in x:\n",
    "                continue\n",
    "            elif 'Mitochondrion' in x:\n",
    "                continue\n",
    "            elif 'non-nuclear' in x:\n",
    "                continue\n",
    "            elif 'secondary' in x:\n",
    "                continue\n",
    "            elif 'Alts' in x:\n",
    "                continue\n",
    "            elif 'total-length' in x:\n",
    "                size.append(x)\n",
    "\n",
    "#Create array of genome size\n",
    "gensize = np.array([706824083, 110781312, 621706043, 481803763, 650477627, 434128924, 389907520, 89583723, 248654244, 1870673364, \n",
    "    191142546, 317690795, 1776217663, 485009472, 249185056, 248577651, 750403944, 225251012, 971188624, \n",
    "    368925819, 398979080, 165928604, 297309692, 90815494, 2738704917, 117319042, 3088623987, 541675471, 56262437, 765382190])\n",
    "\n",
    "#Create dict\n",
    "d = {'Name': names,'Size (bp)':gensize}\n",
    "#print(d)\n",
    "\n",
    "#Create df\n",
    "ncbidf= pd.DataFrame(d).sort_values('Name')\n",
    "\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Data Wrangling -- fixing some abnormal data...ended up being easier to create lists instead of merging dfs\n",
    "\n",
    "Species = norm_insecta_df['Name'].to_list()\n",
    "Order = norm_insecta_df['Classification'].to_list()\n",
    "Genome_Size=ncbidf['Size (bp)'].to_list()\n",
    "Num_of_Genes = norm_GFFdf['Num_of_Genes'].to_list()\n",
    "Gene_Len = norm_GFFdf['Gene_Length'].to_list()\n",
    "Num_of_Exons = norm_GFFdf['Num_of_Exons'].to_list()\n",
    "Exon_Length = norm_GFFdf['Exon_Length'].to_list()\n",
    "Num_of_CDS = norm_GFFdf['Num_of_CDS'].to_list()\n",
    "CDS_Length = norm_GFFdf['CDS_Length'].to_list()\n",
    "\n",
    "\n",
    "final_dict = {\n",
    "                'Species': Species,\n",
    "                'Order': Order,\n",
    "                'Genome Size' : Genome_Size,\n",
    "                'Num of Genes': Num_of_Genes,\n",
    "                'Gene Length': Gene_Len,\n",
    "                'Num of Exons': Num_of_Exons,\n",
    "                'Exon Length': Exon_Length,\n",
    "                'Num of CDS': Num_of_CDS,\n",
    "                'CDS Length': Num_of_CDS\n",
    "                    }\n",
    "\n",
    "\n",
    "final_df=pd.DataFrame(final_dict)\n",
    "\n",
    "final_df.to_csv('final_table.csv')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Notebook1",
   "language": "python",
   "name": "myenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
