{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1484,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## IMPORT PACKAGES\n",
    "\n",
    "import os,gzip,itertools,csv,re,sys,io, Bio\n",
    "import os.path\n",
    "import urllib.request\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import urllib\n",
    "import csv\n",
    "import fastaparser as fp\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio import Entrez\n",
    "from ftplib import FTP\n",
    "import warnings\n",
    "from pathlib import Path\n",
    "import difflib \n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=DeprecationWarning)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1565,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Classification</th>\n",
       "      <th>Taxon ID</th>\n",
       "      <th>Assembly</th>\n",
       "      <th>Accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acyrthosiphon pisum</td>\n",
       "      <td>Hemiptera</td>\n",
       "      <td>7029</td>\n",
       "      <td>Acyr_2.0</td>\n",
       "      <td>GCA_000142985.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Aedes aegypti (LVP_AGWG)</td>\n",
       "      <td>Diptera</td>\n",
       "      <td>7159</td>\n",
       "      <td>AaegL5</td>\n",
       "      <td>GCA_002204515.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Aedes albopictus</td>\n",
       "      <td>Diptera</td>\n",
       "      <td>7160</td>\n",
       "      <td>AaloF1</td>\n",
       "      <td>GCA_001444175.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Apis mellifera (DH4)</td>\n",
       "      <td>Hymenoptera</td>\n",
       "      <td>7460</td>\n",
       "      <td>Amel_HAv3.1</td>\n",
       "      <td>GCA_003254395.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Atta cephalotes</td>\n",
       "      <td>Hymenoptera</td>\n",
       "      <td>12957</td>\n",
       "      <td>Attacep1.0</td>\n",
       "      <td>GCA_000143395.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Belgica antarctica</td>\n",
       "      <td>Diptera</td>\n",
       "      <td>315563</td>\n",
       "      <td>ASM77530v1</td>\n",
       "      <td>GCA_000775305.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Bombus impatiens</td>\n",
       "      <td>Hymenoptera</td>\n",
       "      <td>132113</td>\n",
       "      <td>BIMP_2.0</td>\n",
       "      <td>GCA_000188095.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Bombus terrestris</td>\n",
       "      <td>Hymenoptera</td>\n",
       "      <td>30195</td>\n",
       "      <td>Bter_1.0</td>\n",
       "      <td>GCA_000214255.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Bombyx mori</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>7091</td>\n",
       "      <td>ASM15162v1</td>\n",
       "      <td>GCA_000151625.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Cimex lectularius</td>\n",
       "      <td>Hemiptera</td>\n",
       "      <td>79782</td>\n",
       "      <td>ClecH1</td>\n",
       "      <td>GCA_000648675.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>Danaus plexippus</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>13037</td>\n",
       "      <td>Dpv3</td>\n",
       "      <td>GCA_000235995.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>Heliconius melpomene</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>34740</td>\n",
       "      <td>Hmel1</td>\n",
       "      <td>GCA_000313835.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>Ixodes scapularis</td>\n",
       "      <td>Chelicerata</td>\n",
       "      <td>6945</td>\n",
       "      <td>IscaW1</td>\n",
       "      <td>GCA_000208615.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>Leptotrombidium deliense</td>\n",
       "      <td>Chelicerata</td>\n",
       "      <td>299467</td>\n",
       "      <td>LdelU1</td>\n",
       "      <td>GCA_003675905.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>Lucilia cuprina</td>\n",
       "      <td>Diptera</td>\n",
       "      <td>7375</td>\n",
       "      <td>ASM118794v1</td>\n",
       "      <td>GCA_001187945.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>Melitaea cinxia</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>113334</td>\n",
       "      <td>MelCinx1.0</td>\n",
       "      <td>GCA_000716385.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>Musca domestica</td>\n",
       "      <td>Diptera</td>\n",
       "      <td>7370</td>\n",
       "      <td>MdomA1</td>\n",
       "      <td>GCA_000371365.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>Nasonia vitripennis (AsymCx)</td>\n",
       "      <td>Hymenoptera</td>\n",
       "      <td>7425</td>\n",
       "      <td>Nvit_psr_1.1</td>\n",
       "      <td>GCA_009193385.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>Pediculus humanus</td>\n",
       "      <td>Phthiraptera</td>\n",
       "      <td>121224</td>\n",
       "      <td>PhumU2</td>\n",
       "      <td>GCA_000006295.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>Rhodnius prolixus</td>\n",
       "      <td>Hemiptera</td>\n",
       "      <td>13249</td>\n",
       "      <td>RproC3</td>\n",
       "      <td>GCA_000181055.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>Sarcoptes scabiei</td>\n",
       "      <td>Chelicerata</td>\n",
       "      <td>52283</td>\n",
       "      <td>SscaA1</td>\n",
       "      <td>GCA_000828355.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>Solenopsis invicta</td>\n",
       "      <td>Hymenoptera</td>\n",
       "      <td>13686</td>\n",
       "      <td>Si_gnH</td>\n",
       "      <td>GCA_000188075.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>Stegodyphus mimosarum</td>\n",
       "      <td>Chelicerata</td>\n",
       "      <td>407821</td>\n",
       "      <td>Stegodyphus_mimosarum_v1</td>\n",
       "      <td>GCA_000611955.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>Stomoxys calcitrans</td>\n",
       "      <td>Diptera</td>\n",
       "      <td>35570</td>\n",
       "      <td>ScalU1</td>\n",
       "      <td>GCA_001015335.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>Teleopsis dalmanni</td>\n",
       "      <td>Diptera</td>\n",
       "      <td>139649</td>\n",
       "      <td>Tel_dalmanni_2A_v1.0</td>\n",
       "      <td>GCA_002237135.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>Tetranychus urticae</td>\n",
       "      <td>Chelicerata</td>\n",
       "      <td>32264</td>\n",
       "      <td>ASM23943v1</td>\n",
       "      <td>GCA_000239435.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>Tigriopus californicus (San Diego)</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>6832</td>\n",
       "      <td>Tcal_SD_v2.1</td>\n",
       "      <td>GCA_007210705.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>Tribolium castaneum</td>\n",
       "      <td>Coleoptera</td>\n",
       "      <td>7070</td>\n",
       "      <td>Tcas5.2</td>\n",
       "      <td>GCA_000002335.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>Varroa destructor</td>\n",
       "      <td>Chelicerata</td>\n",
       "      <td>109461</td>\n",
       "      <td>Vdes_3.0</td>\n",
       "      <td>GCA_002443255.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>Zootermopsis nevadensis</td>\n",
       "      <td>Isoptera</td>\n",
       "      <td>136037</td>\n",
       "      <td>ZooNev1.0</td>\n",
       "      <td>GCA_000696155.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   Name Classification  Taxon ID  \\\n",
       "0                   Acyrthosiphon pisum      Hemiptera      7029   \n",
       "2              Aedes aegypti (LVP_AGWG)        Diptera      7159   \n",
       "3                      Aedes albopictus        Diptera      7160   \n",
       "28                 Apis mellifera (DH4)    Hymenoptera      7460   \n",
       "29                      Atta cephalotes    Hymenoptera     12957   \n",
       "30                   Belgica antarctica        Diptera    315563   \n",
       "32                     Bombus impatiens    Hymenoptera    132113   \n",
       "33                    Bombus terrestris    Hymenoptera     30195   \n",
       "34                          Bombyx mori    Lepidoptera      7091   \n",
       "43                    Cimex lectularius      Hemiptera     79782   \n",
       "48                     Danaus plexippus    Lepidoptera     13037   \n",
       "72                 Heliconius melpomene    Lepidoptera     34740   \n",
       "75                    Ixodes scapularis    Chelicerata      6945   \n",
       "78             Leptotrombidium deliense    Chelicerata    299467   \n",
       "82                      Lucilia cuprina        Diptera      7375   \n",
       "86                      Melitaea cinxia    Lepidoptera    113334   \n",
       "88                      Musca domestica        Diptera      7370   \n",
       "89         Nasonia vitripennis (AsymCx)    Hymenoptera      7425   \n",
       "94                    Pediculus humanus   Phthiraptera    121224   \n",
       "97                    Rhodnius prolixus      Hemiptera     13249   \n",
       "98                    Sarcoptes scabiei    Chelicerata     52283   \n",
       "100                  Solenopsis invicta    Hymenoptera     13686   \n",
       "101               Stegodyphus mimosarum    Chelicerata    407821   \n",
       "102                 Stomoxys calcitrans        Diptera     35570   \n",
       "106                  Teleopsis dalmanni        Diptera    139649   \n",
       "107                 Tetranychus urticae    Chelicerata     32264   \n",
       "109  Tigriopus californicus (San Diego)      Crustacea      6832   \n",
       "110                 Tribolium castaneum     Coleoptera      7070   \n",
       "113                   Varroa destructor    Chelicerata    109461   \n",
       "114             Zootermopsis nevadensis       Isoptera    136037   \n",
       "\n",
       "                     Assembly        Accession  \n",
       "0                    Acyr_2.0  GCA_000142985.2  \n",
       "2                      AaegL5  GCA_002204515.1  \n",
       "3                      AaloF1  GCA_001444175.1  \n",
       "28                Amel_HAv3.1  GCA_003254395.2  \n",
       "29                 Attacep1.0  GCA_000143395.2  \n",
       "30                 ASM77530v1  GCA_000775305.1  \n",
       "32                   BIMP_2.0  GCA_000188095.2  \n",
       "33                   Bter_1.0  GCA_000214255.1  \n",
       "34                 ASM15162v1  GCA_000151625.1  \n",
       "43                     ClecH1  GCA_000648675.1  \n",
       "48                       Dpv3  GCA_000235995.2  \n",
       "72                      Hmel1  GCA_000313835.1  \n",
       "75                     IscaW1  GCA_000208615.1  \n",
       "78                     LdelU1  GCA_003675905.1  \n",
       "82                ASM118794v1  GCA_001187945.1  \n",
       "86                 MelCinx1.0  GCA_000716385.1  \n",
       "88                     MdomA1  GCA_000371365.1  \n",
       "89               Nvit_psr_1.1  GCA_009193385.2  \n",
       "94                     PhumU2  GCA_000006295.1  \n",
       "97                     RproC3  GCA_000181055.3  \n",
       "98                     SscaA1  GCA_000828355.1  \n",
       "100                    Si_gnH  GCA_000188075.2  \n",
       "101  Stegodyphus_mimosarum_v1  GCA_000611955.2  \n",
       "102                    ScalU1  GCA_001015335.1  \n",
       "106      Tel_dalmanni_2A_v1.0  GCA_002237135.1  \n",
       "107                ASM23943v1  GCA_000239435.1  \n",
       "109              Tcal_SD_v2.1  GCA_007210705.1  \n",
       "110                   Tcas5.2  GCA_000002335.3  \n",
       "113                  Vdes_3.0  GCA_002443255.1  \n",
       "114                 ZooNev1.0  GCA_000696155.1  "
      ]
     },
     "execution_count": 1565,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## READING IN SUMMARY DATA\n",
    "\n",
    "dat = pd.read_csv('ensemble_metazoa.csv')\n",
    "\n",
    "\n",
    "#with open('ensemble_metazoa.csv', \"r\", newline='') as csvfile:\n",
    "#    read = csv.reader(csvfile, delimiter=',')\n",
    "#    for row in read:\n",
    "#        if row[2] != \"Crustacea\" or \"Hemiptera\" or \"Diptera\" or \"Coleoptera\" or \"Hymenoptera\" or \"Lepidoptera\" or \"Chelicerata\" or \"Phthiraptera\" or \"Isoptera\":\n",
    "            \n",
    "#dat\n",
    "#dat.Classification\n",
    "\n",
    "insecta = dat[(dat.Accession.isin([\"GCA_000142985.2\", \"GCA_002204515.1\", \"GCA_001444175.1\", \"GCA_003254395.2\", \"GCA_000143395.2\", \"GCA_000775305.1\", \"GCA_000188095.2\", \"GCA_000214255.1\", \"GCA_000151625.1\", \"GCA_000648675.1\", \"GCA_000235995.2\", \"GCA_000313835.1\", \"GCA_000208615.1\", \"GCA_002892825.2\", \"GCA_003675905.1\", \"GCA_001187945.1\", \"GCA_000716385.1\", \"GCA_000371365.1\", \"GCA_009193385.2\", \"GCA_000006295.1\", \"GCA_000181055.3\", \"GCA_000828355.1\", \"GCA_000188075.2\", \"GCA_000611955.2\", \"GCA_001015335.1\", \"GCA_002237135.1\", \"GCA_000239435.1\", \"GCA_007210705.1\", \"GCA_000002335.3\", \"GCA_002443255.1\", \"GCA_000696155.1\"]))]\n",
    "\n",
    "insecta_df = insecta[[\"Name\", \"Classification\", \"Taxon ID\", \"Assembly\", \"Accession\"]]\n",
    "\n",
    "#removing duplicates\n",
    "norm_insecta_df = insecta_df.drop(insecta_df.index[13])\n",
    "\n",
    "#norm_insecta_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 624,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## CODE TO PARSE FASTA FILES (Did not use this)\n",
    "# this is code which will parse FASTA files\n",
    "# define what a header looks like in FASTA format\n",
    "def isheader(line):\n",
    "    return line[0] == '>'\n",
    "\n",
    "def aspairs(f):\n",
    "    seq_id = ''\n",
    "    sequence = ''\n",
    "    for header,group in itertools.groupby(f, isheader):\n",
    "        if header:\n",
    "            line = next(group)\n",
    "            seq_id = line[1:].split()[0]\n",
    "        else:\n",
    "            sequence = ''.join(line.strip() for line in group)\n",
    "            yield seq_id, sequence\n",
    "            \n",
    "            \n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1567,
   "metadata": {},
   "outputs": [],
   "source": [
    "## GFF3 DATA PARSING\n",
    "\n",
    "# Get names\n",
    "GFFdirnames = os.listdir('GFF3s')\n",
    "\n",
    "#Loop through files\n",
    "GFFdir = os.scandir('GFF3s')\n",
    "#print(GFFdirnames)\n",
    "\n",
    "## Obtain genome stats\n",
    "GFFdat = []\n",
    "for f in GFFdir:\n",
    "    with gzip.open(f, \"rt\") as fh:\n",
    "        mygff = csv.reader(fh, delimiter=\"\\t\")\n",
    "        count = 0\n",
    "        gene_len = 0\n",
    "        CDS_count = 0\n",
    "        CDS_len = 0\n",
    "        exon_count = 0\n",
    "        exon_len = 0\n",
    "        intron_count = 0\n",
    "        intron_len = 0\n",
    "        genome_size=0\n",
    "        for row in mygff:\n",
    "            if row[0].startswith(\"#\"):\n",
    "                continue\n",
    "            #print(row)\n",
    "            if row[2] == \"gene\":\n",
    "                count += 1\n",
    "                total_gene_len = int(row[4])-int(row[3])\n",
    "                gene_len += total_gene_len\n",
    "            if row[2] == \"CDS\":\n",
    "                CDS_count += 1\n",
    "                total_CDS_len =int(row[4])-int(row[3])\n",
    "                CDS_len +=  total_CDS_len\n",
    "            if row[2] == \"exon\":\n",
    "                exon_count += 1\n",
    "                total_exon_len =int(row[4])-int(row[3])\n",
    "                exon_len +=  total_exon_len\n",
    "                    \n",
    "        mylist=list([Path(f).stem, count, gene_len, exon_count, exon_len, CDS_count, CDS_len])\n",
    "\n",
    "    GFFdat.append(mylist)\n",
    "#print(GFFdat)\n",
    "           "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1572,
   "metadata": {},
   "outputs": [],
   "source": [
    "## CREATE DF FROM GFF DATA\n",
    "\n",
    "GFFdf = pd.DataFrame(GFFdat, columns = [\"Name\",\"Num_of_Genes\",\"Gene_Length\",\"Num_of_Exons\",\"Exon_Length\",\"Num_of_CDS\",\"CDS_Length\"]).sort_values(\"Name\") \n",
    "GFFdf\n",
    "\n",
    "#removing duplicates\n",
    "norm1_GFFdf=GFFdf.drop(GFFdf.index[13])\n",
    "norm_GFFdf=norm1_GFFdf.drop(GFFdf.index[27])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1527,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[541675471, 1870673364, 1776217663, 225251012, 317690795, 89583723, 249185056, 248654244, 481803763, 650477627, 248577651, 3088623987, 765382190, 117319042, 434128924, 389907520, 750403944, 297309692, 110781312, 706824083, 56262437, 398979080, 2738704917, 971188624, 621706043, 90815494, 191142546, 165928604, 368925819, 485009472]\n"
     ]
    }
   ],
   "source": [
    "##OBTAIN GENOME SIZE\n",
    "\n",
    "gensizedir = os.scandir('ncbi-genomes-2020-12-17')\n",
    "\n",
    "names = []\n",
    "size = []\n",
    "for t in gensizedir:\n",
    "    with open(t, \"r\") as txt:\n",
    "        for x in txt:\n",
    "            #print(x)\n",
    "            if \"Organism name\" in x:\n",
    "                names.append(x)\n",
    "            if x.startswith('#'):\n",
    "                continue\n",
    "            elif 'Primary Assembly' in x:\n",
    "                continue\n",
    "            elif 'assembled-molecule' in x:\n",
    "                continue\n",
    "            elif 'Mitochondrion' in x:\n",
    "                continue\n",
    "            elif 'non-nuclear' in x:\n",
    "                continue\n",
    "            elif 'secondary' in x:\n",
    "                continue\n",
    "            elif 'Alts' in x:\n",
    "                continue\n",
    "            elif 'total-length' in x:\n",
    "                size.append(x)\n",
    "\n",
    "#Create array of genome size\n",
    "gensize = np.array([706824083, 110781312, 621706043, 481803763, 650477627, 434128924, 389907520, 89583723, 248654244, 1870673364, \n",
    "    191142546, 317690795, 1776217663, 485009472, 249185056, 248577651, 750403944, 225251012, 971188624, \n",
    "    368925819, 398979080, 165928604, 297309692, 90815494, 2738704917, 117319042, 3088623987, 541675471, 56262437, 765382190])\n",
    "\n",
    "#Create dict\n",
    "d = {'Name': names,'Size (bp)':gensize}\n",
    "#print(d)\n",
    "\n",
    "#Create df\n",
    "ncbidf= pd.DataFrame(d).sort_values('Name')\n",
    "\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1576,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Data Wrangling -- fixing some abnormal data...ended up being easier to create lists instead of merging dfs\n",
    "\n",
    "Species = norm_insecta_df['Name'].to_list()\n",
    "Order = norm_insecta_df['Classification'].to_list()\n",
    "Genome_Size=ncbidf['Size (bp)'].to_list()\n",
    "Num_of_Genes = norm_GFFdf['Num_of_Genes'].to_list()\n",
    "Gene_Len = norm_GFFdf['Gene_Length'].to_list()\n",
    "Num_of_Exons = norm_GFFdf['Num_of_Exons'].to_list()\n",
    "Exon_Length = norm_GFFdf['Exon_Length'].to_list()\n",
    "Num_of_CDS = norm_GFFdf['Num_of_CDS'].to_list()\n",
    "CDS_Length = norm_GFFdf['CDS_Length'].to_list()\n",
    "\n",
    "\n",
    "final_dict = {\n",
    "                'Species': Species,\n",
    "                'Order': Order,\n",
    "                'Genome Size' : Genome_Size,\n",
    "                'Num of Genes': Num_of_Genes,\n",
    "                'Gene Length': Gene_Len,\n",
    "                'Num of Exons': Num_of_Exons,\n",
    "                'Exon Length': Exon_Length,\n",
    "                'Num of CDS': Num_of_CDS,\n",
    "                'CDS Length': Num_of_CDS\n",
    "                    }\n",
    "\n",
    "\n",
    "final_df=pd.DataFrame(final_dict)\n",
    "\n",
    "final_df.to_csv('final_table.csv')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Notebook1",
   "language": "python",
   "name": "myenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
